ls1intum/Artemis
# Create participations for exam exercises if they don't exist

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) **on test server 3**

### Motivation and Context


If a instructor forgets to generate the participations, the participations are generated before starting an exercise on the client

### Steps for Testing


1. Prepare an exam (no modeling exercises), but do not generate the participation for the exercises (don't click prepare exercise start)
2. Participating in exam should work as usual
3. If you are offline when navigating to the exercise for the very first time (not reproducible by reloading, the participation for that exercise must be deleted to have a new try) or the preparation (generation of participation or submission does not work) -> Error + retry button is shown - see screenshot
4. Test for programming exercises (special case)

Known issues which will be tackled in follow up PRs:

1. Retrying the generation of participations might fail, if the client for whatever reason don't get the server response, but the server does prepare the exercise (rare edge case)
2. After trying the generation one time offline, the exercise is not drawn immediately when the retry to generate participations works (changing exercise solves that)
3. Programming exercises can be initiated, but the synchronization behavior does not work as expected until the first submission (look at the navigation bar)

### Screenshots


**Generating Participation**
![image (4)](https://user-images.githubusercontent.com/22955066/86002295-71389c80-ba10-11ea-9c8b-4d1ce70f7baa.png)
**Error occured**
![image](https://user-images.githubusercontent.com/22955066/86002018-07b88e00-ba10-11ea-8816-f457472fae87.png)
